In [33]:
# Import the modules
import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.metrics import balanced_accuracy_score, confusion_matrix, classification_report

import warnings
warnings.filterwarnings('ignore')

---

## Split the Data into Training and Testing Sets

### Step 1: Read the `lending_data.csv` data from the `Resources` folder into a Pandas DataFrame.

In [34]:
# Read the CSV file from the Resources folder into a Pandas DataFrame
# YOUR CODE HERE!
lending_data_df = pd.read_csv(
    "Resources/lending_data.csv")

# Review the DataFrame
# YOUR CODE HERE!
lending_data_df.head()

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  loan_status  
0                 1       22800            0  
1                 0       13600            0  
2                 0       16100            0  
3                 1       22700            0  
4                 1       23000            0

### Step 2: Create the labels set (`y`)  from the “loan_status” column, and then create the features (`X`) DataFrame from the remaining columns.

In [35]:
# Separate the data into labels and features

# Separate the y variable, the labels
# YOUR CODE HERE!]
y = lending_data_df['loan_status']

# Separate the X variable, the features
# YOUR CODE HERE!
X = lending_data_df.drop(columns=['loan_status'])

In [36]:
# Review the y variable Series
# YOUR CODE HERE!
y.head()

0    0
1    0
2    0
3    0
4    0
Name: loan_status, dtype: int64

In [37]:
# Review the X variable DataFrame
# YOUR CODE HERE!
X.head()

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  
0                 1       22800  
1                 0       13600  
2                 0       16100  
3                 1       22700  
4                 1       23000

### Step 3: Check the balance of the labels variable (`y`) by using the `value_counts` function.

In [38]:
# Check the balance of our target values
# YOUR CODE HERE!
y.value_counts()

loan_status
0    75036
1     2500
Name: count, dtype: int64

### Step 4: Split the data into training and testing datasets by using `train_test_split`.

In [39]:
# Import the train_test_learn module
from sklearn.model_selection import train_test_split

# Split the data using train_test_split
# Assign a random_state of 1 to the function
# YOUR CODE HERE!
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1, stratify=y)

---

## Create a Logistic Regression Model with the Original Data

###  Step 1: Fit a logistic regression model by using the training data (`X_train` and `y_train`).

In [40]:
# Import the LogisticRegression module from SKLearn
from sklearn.linear_model import LogisticRegression

# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
# YOUR CODE HERE!
logistic_regression_model = LogisticRegression(random_state=1)

# Fit the model using training data
# YOUR CODE HERE!
lr_model = logistic_regression_model.fit(X_train, y_train)

### Step 2: Save the predictions on the testing data labels by using the testing feature data (`X_test`) and the fitted model.

In [41]:
# Make a prediction using the testing data
# YOUR CODE HERE!
test_predictions = lr_model.predict(X_test)
print(len(test_predictions))

19384


### Step 3: Evaluate the model’s performance by doing the following:

* Calculate the accuracy score of the model.

* Generate a confusion matrix.

* Print the classification report.

In [42]:
# Print the balanced_accuracy score of the model
# YOUR CODE HERE!
accuracy_score= balanced_accuracy_score(y_test, test_predictions)
accuracy_score

0.9442676901753825

In [43]:
# Generate a confusion matrix for the model
# YOUR CODE HERE!
confusion_matrix(y_test, test_predictions)

array([[18679,    80],
       [   67,   558]], dtype=int64)

In [44]:
# Print the classification report for the model
# YOUR CODE HERE!
target_names= ["low_risk_loan", "high_risk_loan"]
print(classification_report(y_test, test_predictions, target_names=target_names))

                precision    recall  f1-score   support

 low_risk_loan       1.00      1.00      1.00     18759
high_risk_loan       0.87      0.89      0.88       625

      accuracy                           0.99     19384
     macro avg       0.94      0.94      0.94     19384
  weighted avg       0.99      0.99      0.99     19384



### Step 4: Answer the following question.

**Question:** How well does the logistic regression model predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** The accuracy over all is 94.4%. For low risk loans the precision is as good as it can be at 1.00. The recall is 1.00 and the f1-score is 1.00. To say the model does well in predicting low risk loans would be an understatement. For high risk loans the precision is 0.87, the recall is 0.89, and the f1-score is 0.88. The model does relatively well in predicting high-risk loans as well. 

---

## Predict a Logistic Regression Model with Resampled Training Data

### Step 1: Use the `RandomOverSampler` module from the imbalanced-learn library to resample the data. Be sure to confirm that the labels have an equal number of data points. 

In [45]:
# Import the RandomOverSampler module form imbalanced-learn
from imblearn.over_sampling import RandomOverSampler

# Instantiate the random oversampler model
# # Assign a random_state parameter of 1 to the model
# YOUR CODE HERE!
rosm = RandomOverSampler(random_state=1)
# Fit the original training data to the random_oversampler model
# YOUR CODE HERE!
X_oversample, y_oversample = rosm.fit_resample(X_train, y_train)

In [46]:
# Count the distinct values of the resampled labels data
# YOUR CODE HERE!
y_oversample.value_counts()

loan_status
0    56277
1    56277
Name: count, dtype: int64

### Step 2: Use the `LogisticRegression` classifier and the resampled data to fit the model and make predictions.

In [49]:
# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
# YOUR CODE HERE!
logistic_regression_oversample_model = LogisticRegression(random_state=1)
# Fit the model using the resampled training data
# YOUR CODE HERE!
logistic_regression_oversample_model.fit(X_oversample, y_oversample)
# Make a prediction using the testing data
# YOUR CODE HERE!
test_predictions_rosm = logistic_regression_oversample_model.predict(X_oversample)

### Step 3: Evaluate the model’s performance by doing the following:

* Calculate the accuracy score of the model.

* Generate a confusion matrix.

* Print the classification report.

In [50]:
# Print the balanced_accuracy score of the model 
# YOUR CODE HERE!
rosm_accuracy = balanced_accuracy_score(y_oversample, test_predictions_rosm)
print(rosm_accuracy)

0.994180571103648


In [52]:
# Generate a confusion matrix for the model
# YOUR CODE HERE!
rosm_confusion = confusion_matrix(y_oversample, test_predictions_rosm)
print(rosm_confusion)

[[55945   332]
 [  323 55954]]


In [53]:
# Print the classification report for the model
# YOUR CODE HERE!

print(classification_report(y_oversample, test_predictions_rosm, target_names=target_names))

                precision    recall  f1-score   support

 low_risk_loan       0.99      0.99      0.99     56277
high_risk_loan       0.99      0.99      0.99     56277

      accuracy                           0.99    112554
     macro avg       0.99      0.99      0.99    112554
  weighted avg       0.99      0.99      0.99    112554



### Step 4: Answer the following question

**Question:** How well does the logistic regression model, fit with oversampled data, predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** Overall the regression model fit with oversampled data is more accurate (99% vs 94%) vs the imbalanced data set, but the precision with the oversampled data is infinitesimally worse fitting the labels of the low risk loans vs the imbalanced data. The precision scores of 0.99 with high risk loans, however is better than that with the imbalanced data set. 